From: Mel Gorman <mgorman@suse.de>
Date: Thu, 16 Jan 2014 09:53:13 +0000
Subject: [PATCH] mm: numa: serialise parallel get_user_page against THP
 migration pmd-locking

References: Automatic NUMA Balancing (fate#315482)
Patch-mainline: Never, required only for pmd split lock backport

The upstream version of the patch "mm: numa: serialise parallel get_user_page
against THP migration" is not aware of the PMD split locking which is
back-ported to SLE12. This patch reconciles the difference.

Signed-off-by: Mel Gorman <mgorman@suse.de>
---
 mm/migrate.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/mm/migrate.c b/mm/migrate.c
index 6693bbe..be81c8f 100644
--- a/mm/migrate.c
+++ b/mm/migrate.c
@@ -1866,13 +1866,13 @@ fail_putback:
 out_fail:
 	count_vm_events(PGMIGRATE_FAIL, HPAGE_PMD_NR);
 out_dropref:
-	spin_lock(&mm->page_table_lock);
+	ptl = pmd_lock(mm, pmd);
 	if (pmd_same(*pmd, entry)) {
 		entry = pmd_mknonnuma(entry);
 		set_pmd_at(mm, mmun_start, pmd, entry);
 		update_mmu_cache_pmd(vma, address, &entry);
 	}
-	spin_unlock(&mm->page_table_lock);
+	spin_unlock(ptl);
 
 out_unlock:
 	unlock_page(page);
